File: Graphs/courses_two.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <functional>

using namespace std;

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, unordered_set<int>> course_prereq{};
        vector<int> ans{};
        unordered_set<int> to_dfs{};

        for(auto& pair : prerequisites) {
            course_prereq[pair[0]].insert(pair[1]);
            // if(course_prereq.find(pair[1]) == course_prereq.cend()) {
            //     course_prereq[pair[1]] = {};
            // }
        }

        for(int i=0; i<numCourses; i++) {
            to_dfs.insert(i);
        }

        function<void(int)> dfs = [&](int course){
            // if(seen.find(x) != seen.cend())
            if(course_prereq[course].empty()) {
                ans.push_back(course);
                to_dfs.erase(course);
                return;
            }
            for(auto& prereq : course_prereq[course]) {
                dfs(prereq);
            }
            ans.push_back(course);
            to_dfs.erase(course);
            return;
        };

        while(!to_dfs.empty()) {
            dfs(*to_dfs.begin());
            if(loop_found) break;
        };

        return ans;
    }
};

File: Graphs/CoursesTwo.js
class Solution {
    findOrder(numCourses, prerequisites) {
        const coursePrereq = new Map();
        const ans = [];
        const toDfs = new Set();
        const hasBeenAdded = new Map();

        for (const pair of prerequisites) {
            if (!coursePrereq.has(pair[0])) {
                coursePrereq.set(pair[0], new Set());
            }
            coursePrereq.get(pair[0]).add(pair[1]);
        }

        for (let i = 0; i < numCourses; i++) {
            toDfs.add(i);
        }

        let loopFound = false;
        const seen = new Set();
        
        const dfs = (course) => {
            if (seen.has(course) || loopFound) {
                loopFound = true;
                return;
            }
            if (!coursePrereq.has(course) || coursePrereq.get(course).size === 0) {
                if (!hasBeenAdded.get(course)) {
                    ans.push(course);
                    hasBeenAdded.set(course, true);
                }
                toDfs.delete(course);
                return;
            }
            seen.add(course);
            for (const prereq of coursePrereq.get(course)) {
                dfs(prereq);
            }
            ans.push(course);
            hasBeenAdded.set(course, true);
            toDfs.delete(course);
            seen.delete(course);
        };

        while (toDfs.size > 0) {
            dfs(toDfs.values().next().value);
            if (loopFound) break;
        }

        return loopFound ? [] : ans;
    }
}

// Example usage:
const solution = new Solution();
const numCourses = 3;
const prerequisites = [[0,1],[0,2],[1,2]];
console.log(solution.findOrder(numCourses, prerequisites));